src/PostgREST/Query/SqlFragment.hs

```diff
@@ -284,10 +284,12 @@ pgFmtApplyAggregate Nothing _ snippet = snippet
 pgFmtApplyAggregate (Just agg) aggCast snippet =
   pgFmtApplyCast aggCast aggregatedSnippet
   where
-    convertAggFunction :: AggregateFunction -> SQL.Snippet
-    -- Convert from e.g. Sum (the data type) to SUM
     convertAggFunction = SQL.sql . BS.map toUpper . BS.pack . show
-    aggregatedSnippet = convertAggFunction agg <> "(" <> snippet <> ")"
+    aggregatedSnippet = case agg of
+     -- TODO: NULLIF(...,'{null}') does not take into consideration a case with a single element with a null value.
+     -- See https://github.com/PostgREST/postgrest/pull/3640#issuecomment-2334996466
+     ArrayAgg -> "COALESCE(NULLIF(array_agg(" <> snippet <> "),'{null}'),'{}')"
```

Comment:
@laurenceisla Could you show an example so I can try and understand this?

Reply:
Sure, let's take this relationship from the tests:

```
factories <--O2M--> processes
```

There's a factory that is not assigned to any process. A simple embedding would look like this:

```bash
curl 'localhost:3000/factories?select=name,processes(*)&id=eq.4'
```
```json
[{"name":"Factory D","processes":[]}]
```

If we didn't implement the `NULLIF(...,'{null}')`. The result of a to-many spread would be like this:

```bash
curl 'localhost:3000/factories?select=name,...processes(process_ids:id,process_names:name)&id=eq.4'
```
```json
[{"name":"Factory D","process_ids":[null]},"process_names":[null]}]
```

This is due to the `LEFT JOIN`. After implementing the `NULLIF(...,'{null}')` condition, the result is expected:

```json
[{"name":"Factory D","process_ids":[]},"process_names":[]}]
```

Now, let's assign that factory to a process. But set the process name as` null`:

```sql
INSERT INTO test.processes(id,name,factory_id) VALUES(9,null,4);
```

The result of the same request would be:

```json
[{"name":"Factory D","process_ids":[9],"process_names":[]}]
```

It DOES have a process now, but the result shows an empty array, which is not correct. It shoul be `"process_names": [null]`, but it's converted due to `NULLIF(...,'{null}')`. It's not common to have a `name` as `null`, but it could be any other column that admits nulls.

Reply:
> It DOES have a process now, but the result shows an empty array, which is not correct. It shoul be "process_names": [null],

I think such a result with `[null]` is "strict", but not correct. It's useless to have that null there for end users. If the array is empty then it doesn't need a `null`.

There's a case which I thought where having a `null` is a must, and it's already handled correctly:

```
-- update this on data.sql
INSERT INTO processes VALUES (3, NULL, 2, 1);

-- and now
curl 'localhost:3000/factories?select=name,...processes(process_id:id,processes_names:name)&name=eq.Factory%20B'

[{"name":"Factory B","process_id":[3,4],"processes_names":[null,"Process B2"]}]
```

Having a `null` in `[null,"Process B2"]` is a must to ensure the order of `process_id`/`processes_names` matches. If it where `["Process B2"]` then that would be a problem.

Reply:
> I think such a result with [null] is "strict", but not correct.

I'd still consider it correct, since this is not an empty array, it's an array with one value that happens to be `null`. It looks better when we add a `count()` in the previous example:

```bash
curl 'localhost:3000/factories?select=name,...processes(process_ids:id,process_names:name,count())&id=eq.4'
```
```json
[{"name":"Factory D","process_ids":[9],"process_names":[],"count":1}]
```

The `"count": 1` confirms that there should be a value. 

> There's a case which I thought where having a `null` is a must, and it's already handled correctly:

Yes, there's no problem there.

Reply:
> The "count": 1 confirms that there should be a value.
> [{"name":"Factory D","process_ids":[9],"process_names":[],"count":1}]

This is true, the `"process_names":[]` needs to be `"process_names":[null]` otherwise it's confusing. We need to clear the TODO and decide now because fixing this later will imply a breaking change on clients.

Reply:
I'd say just document this is a limitation. The count below seems to expensive to check against this case.

Also add a link to https://stackoverflow.com/questions/24155190/postgresql-left-join-json-agg-ignore-remove-null/33153530#33153530 as a comment and mention that `{null}` == `[null]`, but using `[null]` directly makes postgres err with `malformed array literal`.